File: quicksort.java
// Java Quick Sort
// O(n log (n))

void quickSort(int[] arr, int left, int right) {
   int index = partition(arr, left, right);
   if (left < index - 1) { // Sort left half
      quickSort(arr, left, index - 1);
   }
   if (index < right) { // Sort right half
      quickSort(arr, index, right);
   }
}

int partition(int[] arr, int left, int right) {
   int pivot = arr[(left + right) / 2]; // Pick pivot point
   while (left <= right) {
      // Find element on left that should be on right
      while (arr[left] < pivot) left++; 

      // Find element on right that should be on left
      while (arr[right] > pivot) right--;

      // Swap elements and move left and right indices
      if (left <= right) {
         swap(arr, left, right); // swaps elements
         left++;
         right--;
      }
   }
   return left;
}

File: quicksort.js
// JavaScript Quick Sort

const a = [99, 1, 327, 25, 100, 0, 532, 203, 67];

function quickSort(arr) {
    if (arr.length === 0) {
      return [];
    }
  
    let left = [], 
        right = [], 
        pivot = arr[0];
  
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] < pivot) {
            left.push(arr[i]);
        }
        else {
            right.push(arr[i]);
        }
    }
  
    return quickSort(left).concat(pivot, quickSort(right));
}

console.log(quickSort(a));